base/codeinfovalidation.jl

```diff
@@ -0,0 +1,98 @@
+
+`nslots` = the number of unique `SlotNumber`s in `c.code` (as well as `SlotNumber(1)`)
+`nssavals` = the number of unique `SSAValue`s in `c.code`
+
+1:  `length(c.slotflags) != nslots`
+2:  `length(c.slotnames) != nslots`
+3:  `c.inferred && length(c.slottypes) != nslots`
+4:  `c.inferred && length(c.ssavaluetypes) != nssavals`
+5:  `!(c.inferred) && c.slottypes != nothing`
+6:  `!(c.inferred) && c.ssavaluetypes != nssavals`
+7:  `!(in(x.head, Base.Core.VALID_EXPR_HEADS))` for any subexpression `x`
+8:  `length(c.slotnames) < 1 || c.slotnames[1] != Symbol("#self#")`
+9:  `!(isa(x, SSAValue) || isa(x, SlotNumber) || isa(x, GlobalRef))` where `x` is an assignment LHS
+10: `isa(x, Expr) || x.head == :gotoifnot` where `x` is a function call argument
+11:  A slot has an invalid slotflag setting for bit flag 2 (assignment property)
+"""
+function validate_code_info(c::CodeInfo)
+    !(c.inferred) && (c.slottypes != nothing) && return 5
+    (length(c.slotnames) < 1 || c.slotnames[1] != Symbol("#self#")) && return 8
+    slotnums = SlotNumber[]
+    ssavals = SSAValue[]
+    error_code = 0
+    walkast(c.code) do x
+        if isa(x, Expr)
+            if !(in(x.head, VALID_EXPR_HEADS))
+                error_code = 7
+                return true
+            elseif x.head == :(=) && !(is_valid_lhs(x.args[1]))
+                error_code = 9
+                return true
+            elseif x.head == :call && !(all(is_valid_call_arg(i) for i in x.args[2:end]))
+                error_code = 10
+                return true
+            end
+        elseif isa(x, SSAValue) && !(in(x, ssavals))
+            push!(ssavals, x)
+        elseif isa(x, SlotNumber) && !(in(x, slotnums))
+            push!(slotnums, x)
+        end
+        return false
+    end
+    error_code != 0 && return error_code
+    in(SlotNumber(1), slotnums) || push!(slotnums, SlotNumber(1))
+    nslots = length(slotnums)
+    nssavals = length(ssavals)
+    length(c.slotflags) != nslots && return 1
+    length(c.slotnames) != nslots && return 2
+    if c.inferred
+        length(c.slottypes) != nslots  && return 3
+        length(c.ssavaluetypes) != nssavals && return 4
+    else
+        c.ssavaluetypes != nssavals && return 6
+    end
+    error_code = 0
+    walkast(c.code) do x
+        if isa(x, Expr) && x.head == :(=)
+            lhs = x.args[1]
+            if isa(lhs, SlotNumber) && !(is_flag_set(c.slotflags[lhs.id], ASSIGNED_FLAG))
+                error_code = 11
+                return true
+            end
+        end
+        return false
+    end
+    return error_code
+end
+
+function walkast(f, stmts::Array)
+    for stmt in stmts
+        f(stmt) && return true
+        if isa(stmt, Expr)
+            walkast(f, stmt.args) && return true
+        end
+    end
+    return false
+end
+
+is_valid_lhs(lhs) = isa(lhs, SlotNumber) || isa(lhs, SSAValue) || isa(lhs, GlobalRef)
+
+is_valid_call_arg(arg) = !(isa(arg, Expr)) || arg.head != :gotoifnot
```

Comment:
Yeah, this was just a start. I'll add the ones you listed.

Maybe someday we can just replace this function with `isa(arg, SSAValue) || isa(arg, SlotNumber)` 😛